async_service/asyncio.py

```diff
@@ -167,11 +166,16 @@ def __init__(
             )
 
     async def _wait_service_tasks(self) -> None:
+        done = asyncio.Event()
         while True:
-            done, pending = await asyncio.wait(
-                tuple(self._service_task_dag.keys()), return_when=asyncio.ALL_COMPLETED
+            pending_service_tasks = tuple(
+                task for task in self._service_task_dag if not task.done()
             )
-            if all(task.done() for task in self._service_task_dag):
+            if pending_service_tasks:
+                pending_service_tasks[0].add_done_callback(lambda fut: done.set())
```

Comment:
Why do we wait for just the first service task? Would it make sense to add a comment here for that?

Reply:
I'll add a comment.

We have to do this in a loop because each time we `await` it is possible that a new task was added or that an old task completes.  This is the reason for computing `pending_service_tasks` inside the loop.

We only wait for the first one to be done as an optimization.  We *could* wait for all of them but we would still have to go back through the loop and recompute the `pending_service_tasks` to see if there were any new ones that showed up....